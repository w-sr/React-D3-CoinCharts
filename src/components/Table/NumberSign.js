import React from "react";
import PropTypes from "prop-types";

const PLUS_CHAR = "+";
const MINUS_CHAR = "\u2212";

const NumberSign = ({ isPositive }) =>
  isPositive ? <span className="small-font green">{PLUS_CHAR}</span> : <span className="small-font">{MINUS_CHAR}</span>;

NumberSign.propTypes = {
  isPositive: PropTypes.bool.isRequired
};

// Use named export for tests
export { PLUS_CHAR, MINUS_CHAR };

export default NumberSign;
