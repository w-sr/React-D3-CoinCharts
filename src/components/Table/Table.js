import React, { Component } from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

import BigAmount from "../BigAmount";
import Flex from "../Flex";
import { border, color, fontSize, fontWeight, size } from "../../styles/constants";

const StyledTable = styled(Flex)`
  box-sizing: border-box;
  text-align: center;
  padding: ${size.large} 0;
  border-bottom: ${border.border};
`;

const TableCell = styled(Flex)`
  color: ${color.slateDark};
  font-size: ${fontSize.mega};
  width: 100%;

  &:not(:first-child) {
    border-left: ${border.border};
  }
`;

const Label = styled.div`
  color: ${color.coinchartsGray};
  font-size: ${fontSize.small};
  font-weight: ${fontWeight.medium};
  letter-spacing: ${size.tiny};
  text-transform: uppercase;
`;

class Table extends Component {
  shouldComponentUpdate(nextProps) {
    const { isLoading } = nextProps;
    return !isLoading;
  }

  render() {
    const { cryptocurrencyLabel, durationLabel, percentDifference, priceDifference, spotPrice, currency } = this.props;
    const showOtherCells = Boolean(durationLabel);

    return (
      <StyledTable justify="space-around">
        <TableCell column justify="center">
          <BigAmount type="currency" value={spotPrice} currency={currency} />
          <Label>{cryptocurrencyLabel} price</Label>
        </TableCell>
        {showOtherCells && (
          <TableCell column justify="center">
            <BigAmount type="currency" value={priceDifference} currency={currency} showPlusMinusSymbol />
            <Label>{`${durationLabel} (${currency})`}</Label>
          </TableCell>
        )}
        {showOtherCells && (
          <TableCell column justify="center">
            <BigAmount type="percentage" value={percentDifference} showPlusMinusSymbol />
            <Label>{durationLabel} (%)</Label>
          </TableCell>
        )}
      </StyledTable>
    );
  }
}

Table.propTypes = {
  cryptocurrencyLabel: PropTypes.string.isRequired,
  durationLabel: PropTypes.string.isRequired,
  isLoading: PropTypes.bool.isRequired,
  percentDifference: PropTypes.number.isRequired,
  priceDifference: PropTypes.number.isRequired,
  spotPrice: PropTypes.number.isRequired,
  currency: PropTypes.string.isRequired,
};

export default Table;
