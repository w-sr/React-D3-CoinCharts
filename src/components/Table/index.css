.Table {
  box-sizing: border-box;
  display: flex;
  flex-direction: row;
  justify-content: space-around;
}

.TableCell {
  display: flex;
  color: #4e5c6e;
  flex-direction: column;
  font-size: 48px;
  justify-content: center;
  line-height: normal;
  width: 100%;
}

.TableCell:not(:first-child) {
  border-left: 1px solid #dae1e9;
}

.TableCell .label {
  color: #7d95b6;
  font-size: 14px;
  font-weight: 500;
  letter-spacing: 2px;
  text-align: center;
  text-transform: uppercase;
}

.TableCell .value {
  text-align: center;
}

.TableCell .value .green {
  color: #61ca00;
}

.TableCell .value .large-font {
  font-size: 48px;
  font-weight: normal;
  vertical-align: baseline;
}

.TableCell .value .small-font {
  font-size: 30px;
  font-weight: 500;
  top: -13px;
  position: relative;
}
