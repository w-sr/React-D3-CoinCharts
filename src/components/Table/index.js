import React from "react";
import PropTypes from "prop-types";
import { scan } from "d3-array";

import { DEFAULT_PROPS, PROPTYPES } from "../../constants";
import TableCell from "./TableCell";

import "./index.css";

const PriceTable = ({
  cryptocurrencyLabel,
  durationLabel,
  spotPrice,
  priceHistory
}) => {
  const lastIndex = scan(priceHistory, (a, b) => a.time - b.time);
  const oldPrice = priceHistory[lastIndex] && priceHistory[lastIndex].price;
  const priceDifference = spotPrice - oldPrice;
  const percentageDifference = (spotPrice / oldPrice - 1) * 100 || 0;

  return (
    <div className="PriceTable">
      <TableCell
        label={`${cryptocurrencyLabel} price`}
        isCurrency
        value={spotPrice}
      />
      <TableCell
        showPlusCharacter={priceDifference > 0}
        isCurrency
        label={`${durationLabel} (${DEFAULT_PROPS.CURRENCY})`}
        value={priceDifference}
        visible={Boolean(durationLabel)}
      />
      <TableCell
        showPlusCharacter={percentageDifference > 0}
        isPercentage
        label={`${durationLabel} (%)`}
        value={percentageDifference}
        visible={Boolean(durationLabel)}
      />
    </div>
  );
};

PriceTable.propTypes = {
  cryptocurrencyLabel: PropTypes.string.isRequired,
  durationLabel: PropTypes.string.isRequired,
  priceHistory: PROPTYPES.PRICE_DATA.isRequired,
  spotPrice: PropTypes.number.isRequired
};

export default PriceTable;
