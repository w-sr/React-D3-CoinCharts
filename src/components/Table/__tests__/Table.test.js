import React from "react";
import { render } from "react-testing-library";

import Table from "../";

const TEST_PROPS = {
  cryptocurrencyLabel: "testCryptocurrencyLabel",
  durationLabel: "testDurationLabel",
  currency: "USD",
  isLoading: false,
  percentDifference: 0,
  priceDifference: 0,
  spotPrice: 0,
};

describe("<Table />", () => {
  it("renders without crashing", () => {
    const { container } = render(<Table {...TEST_PROPS} />);
    expect(container.firstChild).not.toBeNull();
    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders 3 Table Cells when `props.durationLabel` is defined", () => {
    const { container } = render(<Table {...TEST_PROPS} />);
    expect(container.firstChild.children).toHaveLength(3);
  });

  it("renders only 1 Table Cell when `props.durationLabel` is undefined", () => {
    const { container } = render(<Table {...TEST_PROPS} durationLabel="" />);
    expect(container.firstChild.children).toHaveLength(1);
  });

  it("does not rerender when `props.isLoading` is true", () => {
    const { container } = render(<Table {...TEST_PROPS} />);
    const { textContent } = container.firstChild;

    // Component should not rerender
    render(<Table {...TEST_PROPS} percentDifference={1000} isLoading />, { container });
    expect(container.firstChild.textContent).toEqual(textContent);

    // Component should rerender
    render(<Table {...TEST_PROPS} percentDifference={1000} />, { container });
    expect(container.firstChild.textContent).not.toEqual(textContent);
  });
});
