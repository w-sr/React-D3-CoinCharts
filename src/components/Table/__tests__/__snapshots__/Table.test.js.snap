// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Table /> renders without crashing 1`] = `
<div
  class="sc-bxivhb kNEjoy sc-htpNat bJzcAl"
>
  <div
    class="sc-ifAKCX fhHsZd sc-htpNat Ldmto"
  >
    <div
      class="sc-bdVaJa jlxkRb"
    >
      <span
        color="#4E5C6E"
        class="sc-bwzfXH fbCOIX"
      >
        $
      </span>
      <span>
        0
      </span>
      <span
        color="#4E5C6E"
        class="sc-bwzfXH fbCOIX"
      >
        .00
      </span>
    </div>
    <div
      class="sc-EHOje eDVoFS"
    >
      testCryptocurrencyLabel
      price
    </div>
  </div>
  <div
    class="sc-ifAKCX fhHsZd sc-htpNat Ldmto"
  >
    <div
      class="sc-bdVaJa jlxkRb"
    >
      <span
        color="#4E5C6E"
        class="sc-bwzfXH fbCOIX"
      >
        $
      </span>
      <span>
        0
      </span>
      <span
        color="#4E5C6E"
        class="sc-bwzfXH fbCOIX"
      >
        .00
      </span>
    </div>
    <div
      class="sc-EHOje eDVoFS"
    >
      testDurationLabel (USD)
    </div>
  </div>
  <div
    class="sc-ifAKCX fhHsZd sc-htpNat Ldmto"
  >
    <div
      class="sc-bdVaJa jlxkRb"
    >
      <span>
        0.00
      </span>
      <span
        color="#4E5C6E"
        class="sc-bwzfXH fbCOIX"
      >
        %
      </span>
    </div>
    <div
      class="sc-EHOje eDVoFS"
    >
      testDurationLabel
      (%)
    </div>
  </div>
</div>
`;
